### ISSUE TITLE
Problem: Servo motion not smooth

### ISSUE THREAD

**Issue by MarkJB** [bdring/FluidNC#852]
### Controller Board

ESP32 Plotter Controller Rev B (aka "Emerald Dingo") https://github.com/theworkisthework/ESP32_Plotter_Controller

Mostly pin compatible with [Grbl_ESP32 CNC Development Board](https://www.tindie.com/products/33366583/grbl_esp32-cnc-development-board-v41/)

### Machine Description

At this point in time, I'm just bench testing the DMC809 coreless servo with the controller and no machine.

### Configuration file

```yaml
name: ESP32_Plotter_Controller_RevB_Emerald_Dingo
board: Emerald Dingo (RevB)

kinematics:
  CoreXY:

stepping:
  engine: RMT
  idle_ms: 255
  dir_delay_us: 0
  pulse_us: 3
  disable_delay_us: 0

axes:
  shared_stepper_disable_pin: gpio.2
  x:
    steps_per_mm: 100.000
    max_rate_mm_per_min: 1000.000
    acceleration_mm_per_sec2: 200.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      mpos_mm: 0.000
      positive_direction: true
      settle_ms: 250.000
      seek_mm_per_min: 2000.000
      feed_mm_per_min: 200.000
      seek_scaler: 1.100
      feed_scaler: 5.000
    motor0:
      limit_all_pin: gpio.34:low
      hard_limits: false
      tmc_2130:
        direction_pin: gpio.26
        step_pin: gpio.25
        run_amps: 0.250
        hold_amps: 0.125
        microsteps: 16
        stallguard: 16
        stallguard_debug: false
        run_mode: CoolStep
        homing_mode: CoolStep
        use_enable: false
        r_sense_ohms: 0.110
        cs_pin: gpio.17
    motor1:
      null_motor:
  y:
    steps_per_mm: 100.000
    max_rate_mm_per_min: 1000.000
    acceleration_mm_per_sec2: 200.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      mpos_mm: 0.000
      positive_direction: true
      settle_ms: 250.000
      seek_mm_per_min: 2000.000
      feed_mm_per_min: 200.000
      seek_scaler: 1.100
      feed_scaler: 5.000
    motor0:
      limit_all_pin: gpio.35:low
      hard_limits: false
      tmc_2130:
        direction_pin: gpio.33
        step_pin: gpio.32
        run_amps: 0.250
        hold_amps: 0.125
        microsteps: 16
        stallguard: 16
        stallguard_debug: false
        run_mode: CoolStep
        homing_mode: CoolStep
        use_enable: false
        r_sense_ohms: 0.110
        cs_pin: gpio.16
    motor1:
      null_motor:
  z:
    steps_per_mm: 200.000
    max_rate_mm_per_min: 31000.000
    # max_rate_mm_per_min: 555.000
    acceleration_mm_per_sec2: 20000.000
    max_travel_mm: 24.000
    # KPOWER DMC809 - calculated ~180deg circumference travel = 37mm is more like 24mm linear
    soft_limits: false
    hard_limits: false
    homing:
      cycle: 1
      mpos_mm: 12.000
      positive_direction: true
      settle_ms: 250.000
      seek_mm_per_min: 2000.000
      feed_mm_per_min: 200.000
      seek_scaler: 1.100
      feed_scaler: 5.000
    motor0:
      rc_servo:
        pwm_hz: 50
        output_pin: gpio.27
        min_pulse_us: 2500
        max_pulse_us: 750
    motor1:
      null_motor:

spi:
  miso_pin: gpio.19
  mosi_pin: gpio.23
  sck_pin: gpio.18

sdcard:
  cs_pin: gpio.5

control:
  reset_pin: gpio.0:low:pu

coolant:
  delay_ms: 1000.000

probe:
  check_mode_start: false

macros:
  startup_line0:
  startup_line1:
  macro0:
  macro1:
  macro2:
  macro3:

start:
  must_home: true
  check_limits: true
  deactivate_parking: false

user_outputs:

# relay:
#   tool_num: 0
#   speed_map: 0=0.0% 1000=100.0%
#   spinup_ms: 0
#   spindown_ms: 0
#   output_pin: gpio.27
#   disable_with_s0: false
#   s0_with_disable: false
#   spinup_ms: 0
#   spindown_ms: 0
arc_tolerance_mm: 0.002
junction_deviation_mm: 0.010
verbose_errors: false
report_inches: false
enable_parking_override_control: false
use_line_numbers: false
```

### Startup Messages

```text
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:12812
load:0x40080400,len:3032
entry 0x400805e4

[MSG:INFO: FluidNC v3.6.8]
[MSG:INFO: Compiled with ESP32 SDK:v4.4.1-1-gb8050b365e]
[MSG:INFO: Local filesystem type is spiffs]
[MSG:INFO: Configuration file:config.yaml]
[MSG:WARN: Ignored key hard_limits]
[MSG:INFO: Machine ESP32_Plotter_Controller_RevB_Emerald_Dingo]
[MSG:INFO: Board Emerald Dingo (RevB)]
[MSG:INFO: SPI SCK:gpio.18 MOSI:gpio.23 MISO:gpio.19]
[MSG:INFO: SD Card cs_pin:gpio.5 detect:NO_PIN freq:0]
[MSG:INFO: Stepping:RMT Pulse:3us Dsbl Delay:0us Dir Delay:0us Idle Delay:255ms]
[MSG:INFO: Axis count 3]
[MSG:INFO: Shared stepper disable gpio.2]
[MSG:INFO: Axis X (-300.000,0.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     tmc_2130 Step:gpio.25 Dir:gpio.26 CS:gpio.17 Disable:NO_PIN Index:-1 R:0.110]
[MSG:INFO:  X All Limit gpio.34:low]
[MSG:INFO:   Motor1]
[MSG:INFO: Axis Y (-300.000,0.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     tmc_2130 Step:gpio.32 Dir:gpio.33 CS:gpio.16 Disable:NO_PIN Index:-1 R:0.110]
[MSG:INFO:  Y All Limit gpio.35:low]
[MSG:INFO:   Motor1]
[MSG:INFO: Axis Z (-12.000,12.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     rc_servo Pin:gpio.27 Pulse Len(2500,750 period:1048575)]
[MSG:INFO:   Motor1]
[MSG:INFO: X Axis driver test passed]
[MSG:INFO: Y Axis driver test passed]
[MSG:INFO: reset_pin gpio.0:low:pu]
[MSG:INFO: Kinematic system: CoreXY]
[MSG:INFO: Using spindle NoSpindle]
[MSG:INFO: Connecting to STA SSID:TP-Link_EAC7]
[MSG:INFO: Connecting.]
[MSG:INFO: Connecting..]
[MSG:INFO: Connected - IP is 192.168.1.116]
[MSG:INFO: WiFi on]
[MSG:INFO: Start mDNS with hostname:http://fluidnc.local/]
[MSG:INFO: SSDP Started]
[MSG:INFO: HTTP started on port 80]
[MSG:INFO: Telnet started on port 23]

Grbl 3.6 [FluidNC v3.6.8 (wifi) '$' for help]
[MSG:INFO: '$H'|'$X' to unlock]
[MSG:INFO: Channel auto report interval set to 50 ms]
[MSG:INFO: Caution: Unlocked]
```

### User Interface Software

WebUI, FluidTerm

### What happened?

When using a servo as the Z axis, the servo makes rough moves when using the WebUI Z controls or sending `G0Z20`/`G0Z0`. Was expecting it to be smooth (its smooth on the same controller with GRBLESP). 

It seems like the PWM values are sent in discreet steps (the steps are visible when looking on an oscilisope). The servo I am using is a DMC809 coreless and its very fast. So much so that its getting to the desired position long before the next PWM step change. So its is accelerating fast and stopping abruptly and doing that again when the PWM value changes.

Here is a link to a video of the servo moving: https://photos.app.goo.gl/7kFAoVn3vFfe6hpG6

Also the crazy 'z' 'speeds and feeds' values currently in the config file, are the result of me attempting to find values that work and experimenting way beyond what I thought was likely. These values started low, but nothing seemed to improve the situation so I started putting in wild values on the off chance it gave me some insight into the problem.

**Comment by bdring** [comment]
The speeds and acceleration seem fast. If you assume the servo can do the full travel in 1 second. That is 24mm/sec or 1440 mm/min. If you assume it can reach full speed in 1/2 second then the acceleration is 48 mm_per_sec2.

Try using these values.

Use G53 G1 moves with feed rates and your max travel, like... 
```gcode
G53 G1 Z24 F1400
G53 G1 Z0 F500
```

**Comment by MarkJB** [comment]
The values in the config file are way higher than intended and not realistic. I was working my way up to see if the stuttering was a result of the slower speeds, but using the commands you posted above, the servo moves slower but still stutters. If I view the PWM signal on an oscilloscope I can see it jumping from one pulse width to the next, matching the servos stutters, rather than a smooth transition.

**Comment by MitchBradley** [comment]
The task that updates the servo runs every 75 ms by default.  You can override that by adding:
```
         rc_servo:
             timer_ms: 20
```
Going below 20 will not help because RC servos inherently run at 50 Hz, i.e. 20 ms.

**Comment by MarkJB** [comment]
> The task that updates the servo runs every 75 ms by default. You can override that by adding:
> 
> ```
>          rc_servo:
>              timer_ms: 20
> ```
> 
> Going below 20 will not help because RC servos inherently run at 50 Hz, i.e. 20 ms.

Is there an upper limit for that value?

I wondered if setting it high (like 250ms) would result in a single bigger step allowing the servo to run at its max speed, but it doesn't seem to be changing the frequency of the servos' stutter. 

I then set it to 2500 to see if I could see an obvious difference, but I can't. I'm not sure its having any effect.

Set back to 20ms for testing.

> The speeds and acceleration seem fast. If you assume the servo can do the full travel in 1 second. That is 24mm/sec or 1440 mm/min. If you assume it can reach full speed in 1/2 second then the acceleration is 48 mm_per_sec2.

The servo specs give this:

`Speed: 0.08 sec / 60° / 4.8 V`, so assuming that is accurate, ~120° in 0.16s. ~150mm/sec or 9000mm/min.  No idea on the acceleration, but its fast. So guessed at 250 mm_per_sec2. 

I then ran the following gcode: 
https://gist.github.com/MarkJB/64f7c1ec3eedaae071b3049038b9a090

This seemed to max out so I bumped up the acceleration to 500, then 1000, then 1500 which seem like crazy high values, but also seem to make a difference to the servo rate. (Not verified yet, I need to take some more slow mo videos and work out the actual speed).

This video shows the result: https://photos.app.goo.gl/dGfzx7Rg8ubsr2Xp8

In summary; slow (<F100) is ok and very fast is ok (>F5000 ? maybe - not sure yet). But values between stutter...

Current Z axis config:

```
  z:
    steps_per_mm: 200.000
    max_rate_mm_per_min: 9000.000
    acceleration_mm_per_sec2: 1500.000
    max_travel_mm: 24.000
    soft_limits: false
    hard_limits: false
    homing:
      cycle: 1
      mpos_mm: 12.000
      positive_direction: true
      settle_ms: 250.000
      seek_mm_per_min: 2000.000
      feed_mm_per_min: 200.000
      seek_scaler: 1.100
      feed_scaler: 5.000
    motor0:
      rc_servo:
        pwm_hz: 50
        timer_ms: 20
        output_pin: gpio.27
        min_pulse_us: 2500
        max_pulse_us: 750
    motor1:
      null_motor:

**Comment by bdring** [comment]
Use 20ms and 50Hz.  Start with slow max speeds and accelerations. You can slowly increase those until the servo can't keep up.

Running gcode on a servo is a hack. There is no feedback from the servo if you are exceeding any values.

Here are the ranges

```cpp

const int      SERVO_PWM_FREQ_DEFAULT = 50;  // 50Hz ...This is a standard analog servo value. Digital ones can repeat faster
const uint32_t SERVO_PWM_FREQ_MIN     = 50;
const uint32_t SERVO_PWM_FREQ_MAX     = 200;

const int      SERVO_PULSE_US_MIN_DEFAULT = 1000;
const int      SERVO_PULSE_US_MAX_DEFAULT = 2000;
const uint32_t SERVO_PULSE_US_MIN         = 500;
const uint32_t SERVO_PULSE_US_MAX         = 2500;
```

**Comment by MarkJB** [comment]
I went back to much lower values and tried increasing the values again. Servo still isn't smooth.

I didn't think the `timer_ms` setting was doing anything and eventually noticed `[MSG:WARN: Ignored key timer_ms]` in the console startup messages.

If I set the `_time_ms` variable in `src/Motors/Servo.h` to `20` (defaults to `75`) the servo on the resulting compiled firmware is smooth as silk.

**Comment by bdring** [comment]
Please donate to this project.

**Comment by MarkJB** [comment]
Any idea why the config value is ignored? Or should I raise a separate issue for that?

**Comment by bdring** [comment]
Please post the current config and startup messages if you want me to take some time to figure out your problem

**Comment by MarkJB** [comment]
```
name: ESP32_Plotter_Controller_RevB_Emerald_Dingo
board: Emerald Dingo (RevB)

kinematics:
  CoreXY:

stepping:
  engine: RMT
  idle_ms: 255
  dir_delay_us: 0
  pulse_us: 3
  disable_delay_us: 0

axes:
  shared_stepper_disable_pin: gpio.2
  x:
    steps_per_mm: 100.000
    max_rate_mm_per_min: 1000.000
    acceleration_mm_per_sec2: 200.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      mpos_mm: 0.000
      positive_direction: true
      settle_ms: 250.000
      seek_mm_per_min: 2000.000
      feed_mm_per_min: 200.000
      seek_scaler: 1.100
      feed_scaler: 5.000
    motor0:
      limit_all_pin: gpio.34:low
      hard_limits: false
      tmc_2130:
        direction_pin: gpio.26
        step_pin: gpio.25
        run_amps: 0.250
        hold_amps: 0.125
        microsteps: 16
        stallguard: 16
        stallguard_debug: false
        run_mode: CoolStep
        homing_mode: CoolStep
        use_enable: false
        r_sense_ohms: 0.110
        cs_pin: gpio.17
    motor1:
      null_motor:
  y:
    steps_per_mm: 100.000
    max_rate_mm_per_min: 1000.000
    acceleration_mm_per_sec2: 200.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      mpos_mm: 0.000
      positive_direction: true
      settle_ms: 250.000
      seek_mm_per_min: 2000.000
      feed_mm_per_min: 200.000
      seek_scaler: 1.100
      feed_scaler: 5.000
    motor0:
      limit_all_pin: gpio.35:low
      hard_limits: false
      tmc_2130:
        direction_pin: gpio.33
        step_pin: gpio.32
        run_amps: 0.250
        hold_amps: 0.125
        microsteps: 16
        stallguard: 16
        stallguard_debug: false
        run_mode: CoolStep
        homing_mode: CoolStep
        use_enable: false
        r_sense_ohms: 0.110
        cs_pin: gpio.16
    motor1:
      null_motor:
  z:
    steps_per_mm: 200.000
    max_rate_mm_per_min: 1500.000
    acceleration_mm_per_sec2: 50.000
    max_travel_mm: 24.000
    soft_limits: false
    hard_limits: false
    homing:
      cycle: 1
      mpos_mm: 12.000
      positive_direction: true
      settle_ms: 250.000
      seek_mm_per_min: 2000.000
      feed_mm_per_min: 200.000
      seek_scaler: 1.100
      feed_scaler: 5.000
    motor0:
      rc_servo:
        pwm_hz: 50
        timer_ms: 20
        output_pin: gpio.27
        min_pulse_us: 2500
        max_pulse_us: 750
    motor1:
      null_motor:

spi:
  miso_pin: gpio.19
  mosi_pin: gpio.23
  sck_pin: gpio.18

sdcard:
  cs_pin: gpio.5

control:
  reset_pin: gpio.0:low:pu

coolant:
  delay_ms: 1000.000

probe:
  check_mode_start: false

macros:
  startup_line0:
  startup_line1:
  macro0:
  macro1:
  macro2:
  macro3:

start:
  must_home: true
  check_limits: true
  deactivate_parking: false

user_outputs:

# relay:
#   tool_num: 0
#   speed_map: 0=0.0% 1000=100.0%
#   spinup_ms: 0
#   spindown_ms: 0
#   output_pin: gpio.27
#   disable_with_s0: false
#   s0_with_disable: false
#   spinup_ms: 0
#   spindown_ms: 0
arc_tolerance_mm: 0.002
junction_deviation_mm: 0.010
verbose_errors: false
report_inches: false
enable_parking_override_control: false
use_line_numbers: false
``` 

```
rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:12812
load:0x40080400,len:3032
entry 0x400805e4

[MSG:INFO: FluidNC v3.6.8]
[MSG:INFO: Compiled with ESP32 SDK:v4.4.1-1-gb8050b365e]
[MSG:INFO: Local filesystem type is spiffs]
[MSG:INFO: Configuration file:config.yaml]
[MSG:WARN: Ignored key hard_limits]
[MSG:WARN: Ignored key timer_ms]
[MSG:INFO: Machine ESP32_Plotter_Controller_RevB_Emerald_Dingo]
[MSG:INFO: Board Emerald Dingo (RevB)]
[MSG:INFO: SPI SCK:gpio.18 MOSI:gpio.23 MISO:gpio.19]
[MSG:INFO: SD Card cs_pin:gpio.5 detect:NO_PIN freq:0]
[MSG:INFO: Stepping:RMT Pulse:3us Dsbl Delay:0us Dir Delay:0us Idle Delay:255ms]
[MSG:INFO: Axis count 3]
[MSG:INFO: Shared stepper disable gpio.2]
[MSG:INFO: Axis X (-300.000,0.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     tmc_2130 Step:gpio.25 Dir:gpio.26 CS:gpio.17 Disable:NO_PIN Index:-1 R:0.110]
[MSG:INFO:  X All Limit gpio.34:low]
[MSG:INFO:   Motor1]
[MSG:INFO: Axis Y (-300.000,0.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     tmc_2130 Step:gpio.32 Dir:gpio.33 CS:gpio.16 Disable:NO_PIN Index:-1 R:0.110]
[MSG:INFO:  Y All Limit gpio.35:low]
[MSG:INFO:   Motor1]
[MSG:INFO: Axis Z (-12.000,12.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     rc_servo Pin:gpio.27 Pulse Len(2500,750 period:1048575)]
[MSG:INFO:   Motor1]
[MSG:INFO: X Axis driver test passed]
[MSG:INFO: Y Axis driver test passed]
[MSG:INFO: reset_pin gpio.0:low:pu]
[MSG:INFO: Kinematic system: CoreXY]
[MSG:INFO: Using spindle NoSpindle]
[MSG:INFO: Connecting to STA SSID:TP-Link_EAC7]
[MSG:INFO: Connecting.]
[MSG:INFO: Connecting..]
[MSG:INFO: Connected - IP is 192.168.1.103]
[MSG:INFO: WiFi on]
[MSG:INFO: Start mDNS with hostname:http://fluidnc.local/]
[MSG:INFO: SSDP Started]
[MSG:INFO: HTTP started on port 80]
[MSG:INFO: Telnet started on port 23]

Grbl 3.6 [FluidNC v3.6.8 (wifi) '$' for help]
[MSG:INFO: '$H'|'$X' to unlock]
[MSG:INFO: Channel auto report interval set to 50 ms]
```

**Comment by MarkJB** [comment]
I had a go a fixing the issue - https://github.com/bdring/FluidNC/pull/855

If the PR is valid, I'm happy to update the wiki if I can get access.
